### ISSUE TITLE
more control flow for Rune

### ISSUE THREAD

**Issue by nythrox** [paritytech/capi#828]
Handle isn't enough for to translate JS flows into Rune without being forced to use the `.unhandle .rehandle` pattern

#### Necessary control flows

The `handle` function is great for branching your code based on a condition, but right now it doesn't cover all the very basic control flows JS offers. I'd like it to cover all use cases: if/elseif/else, , exhaustive switch (#624), and error handling. We could also add some new operators to make common flows easier: `&&` `??` `value!` `?.`

Why we need more conditionals: 
`handle()` is not enough to avoid using `.handle()` and `.unhandle()`, since in some of the most common scenarios that this pattern was needed (for a lack of alternative) was when the user would be trying to reproduce a "if true do x, else do y".

```ts
// vanilla TS
const settings = props?.mintSettings?.defaultItemSettings
const defaultItemSettings = settings ? $settings.encode(settings) : 0n

// Rune 
const settings = props.access("mintSettings").access("defaultItemSettings")
const defaultItemSettings = settings
  .unhandle(undefined)
  .chain($settings.encode)
  .rehandle(undefined)
  .handle(undefined, () => 0n)
```

If we were to try to use only `handle` (without the `.unhandle .rehandle` pattern): 
```ts
const defaultItemSettings = settings.handle(undefined, () => 0n).chain($settings.encode)
```
This would give a type error, since after `handle` the type of the value is no longer `Settings | undefined`, it becomes `bigint | Settings` which cannot be used on $settings.encode. This could easily be fixed with some kind of `else` method:
```ts
const defaultItemSettings = settings.handle(undefined, () => 0n, /* else */ $settings.encode) // typechecks correctly as bigint
```

**Comment by nythrox** [comment]
I would suggest a more advanced handle/match:

```ts
settings.handle(undefined => ...)
settings.match(
  handle((v) => ...) // if
  handle((v) => ...) // else if
  otherwise((e) => ...) // else
)
```

and maybe some helpers for operator shortcuts
```ts
settings.handle(truthy, () => ...) // &&
settings.handle(falsy, () => ...) // ||
settings.access("a", "b") // settings?.a?.b
```

I'm not sure about the names, the format, or even satisfied on how it looks, but I still think it is essential for rune to be able to express all the conditional control flows used in JS in more intuitive way than `unhandle rehandle`

**Comment by nythrox** [comment]
Note: this is mostly important pattern lib developers who need to build their program completely inside capi, normal users wouldn't need to use these

**Comment by harrysolovay** [comment]
> The `handle` function [...] doesn't cover all the very basic control flows JS offer

I don't think an exhaustive match / version of handle is a key requirement atm considering that the result of a `handle` call is a rune, which will usually be supplied to another rune factory (and therefore type-checked against the given parameter).

> since in some of the most common scenarios...

Can you please provide some specific examples from our domain (ideally from FRAME pallets). This way we'd have something specific to contemplate/design.

> // vanilla TS
> const settings = props?.mintSettings?.defaultItemSettings
> const defaultItemSettings = settings ? $settings.encode(settings) : 0n

You're right... this is a bit of a pain point at the moment. Hopefully we can get `rune.access` to respect optionals so that we could do the following.

```ts
const settings = Rune
  .resolve(props)
  .access("mintSettings", "defaultItemSettings")
  .handle(undefined, () => Rune.resolve(0n))
```

**Comment by nythrox** [comment]
> Can you please provide some specific examples from our domain (ideally from FRAME pallets). This way we'd have something specific to contemplate/design.

> Hopefully we can get rune.access to respect optionals so that we could do the following.

The initial example is a use case where the problem is not accessing optionals, but using a if/else flow (which currently can only be represented in `.unhandle .rehandle`

Being able to `.access("mintSettings", "defaultItemSettings").handle(undefined, () => Rune.resolve(0n))` would not solve this example, since what we want to do is apply `$settings.encode` **if** the `Settings` object is present, and if not then return a `0n` `bigint` default

The reason why this is a common problem is that it doesn't appear only when doing something based on if something is undefined or not, but in **any** case where you want to do something based on a condition (you could replace `settings` the first example with any variable that you'd want to treat differently based its value

**Comment by harrysolovay** [comment]
> Being able to `.access("mintSettings", "defaultItemSettings").handle(undefined, () => Rune.resolve(0n))` would not solve this example, since what we want to do is apply `$settings.encode` if the `Settings` object is present

Ah, yes: seems I glossed over that. How's this then?:

```ts
const $settings = CodecRune.from($c)
const settings = Rune
  .resolve(props)
  .access("mintSettings", "defaultItemSettings")
  .handle(not(undefined), $settings.decoded)
  .handle(undefined, () => Rune.resolve(0n))
```

We don't have the `not`/guard-building helpers as of yet, but it should be simple enough. I'd imagine that––in most cases––devs will want a given Rune to resolve to a somewhat consistent shape. Too much branching shouldn't usually happen from Runes higher in the tree. But **who knows**?! For cases where one wants to represent such Runes, I still think `handle` and guards will be sufficient. If you disagree, please follow up on my earlier question:

> Can you please provide some specific examples from our domain (ideally from FRAME pallets). This way we'd have something specific to contemplate/design.

**Comment by tjjfvi** [comment]
We'll likely solve this in the same way as #624

**Comment by harrysolovay** [comment]
Resolved as of #1078
